src/plugins/path/path.c

```diff
@@ -65,6 +100,149 @@ static int validateKey (Key * key, Key * parentKey)
+{
+	uid_t currentUID = geteuid ();
+	gid_t currentGID = getegid ();
+
+	const Key * userMeta = keyGetMeta (key, "check/permission/user");
+	const Key * userTypes = keyGetMeta (key, "check/permission/types");
+
+	// ***** central variables *******
+	const char * validPath = keyString (key);
+	const char * name = keyString (userMeta);
+	const char * modes = keyString (userTypes);
+	// ****************************
+
+	struct passwd * p;
+
+	// Changing to specified user. Can only be done when executing user is root user
+	if (userMeta)
+	{
+		p = getpwnam (name);
+		// Check if user exists
+		if (p == NULL)
+		{
+			ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_USER_NOT_FOUND, parentKey,
+					    "Could not find user \"%s\" for key \"%s\". "
+					    "Does the user exist?\"",
+					    name, keyName (key));
+			return -1;
+		}
+		name = p->pw_name;
+
+		// Check if I can change the UID as root
+		int err = seteuid ((int) p->pw_uid);
+		if (err < 0)
+		{
+			ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_CHANGE_USER_ERROR, parentKey,
+					    "Could not set euid of user \"%s\" for key \"%s\"."
+					    " Are you running kdb as root?\"",
+					    name, keyName (key));
+			return -1;
+		}
+	}
+	else
+	{
+		uid_t uid = geteuid ();
+		p = getpwuid (uid);
+		name = p->pw_name;
+		if (uid != 0)
+		{
+			ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_CHANGE_USER_ERROR, parentKey,
+					    "To check permissions for %s I need to be the root user."
+					    " Are you running kdb as root?\"",
+					    keyName (key));
+			return -1;
+		}
+	}
+
+	// The following code changes the egid if a group from a user matches the filegroup
+	// TODO: Whats a good default value for ngroups
+	// If the relevant group is ngroups+1 then it wont get recognized
+	int ngroups = 30;
+	gid_t * groups;
+	groups = (gid_t *) elektraMalloc (ngroups * sizeof (gid_t));
+	getgrouplist (name, (int) p->pw_gid, groups, &ngroups);
+
+	// Get groupID of file being checked
+	struct stat sb;
+	stat (validPath, &sb);
+	struct group * gr = getgrgid (sb.st_gid);
+
+	bool isUserInGroupBool = isUserInGroup ((int) gr->gr_gid, groups, (unsigned int) ngroups);
+
+	// Check if fileGroup is in userGroup. If yes change egid to that group
+	if (isUserInGroupBool)
+	{
+		ELEKTRA_LOG_DEBUG ("User “%s” has group of file %s", name, validPath);
+		int gidErr = setegid ((int) gr->gr_gid);
+		if (gidErr < 0)
+		{
+			ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_CHANGE_USER_ERROR, parentKey,
+					    "Could not set egid of user \"%s\" for key \"%s\"."
+					    " Are you running kdb as root?\"",
+					    name, keyName (key));
+		}
+	}
+	elektraFree (groups);
+
+	// Actual checks are done
+	int isRead = (strchr (modes, 'r') == NULL) ? 0 : 1;
+	int isWrite = (strchr (modes, 'w') == NULL) ? 0 : 1;
+	int isExecute = (strchr (modes, 'x') == NULL) ? 0 : 1;
+
+	char errorMessage[30];
+	errorMessage[0] = '\0'; // strcat() searches for this, otherwise it will print garbage chars at start
+	int isError = 0;
+
+	if (isRead && euidaccess (validPath, R_OK) != 0)
```

Comment:
I was also considering to do that but if I do not do it this way I cannot know which permission the user actually is missing because errno in [euidaccess](http://man7.org/linux/man-pages/man2/access.2.html) does not yield which specific permission was denied

Reply:
I was also considering to do that but if I do not do it this way I cannot know which permission the user actually is missing because errno in euidaccess does not yield which specific permission was denied

Reply:
Do you need to know? I think the error message `"Missing permissions. Required: ..."`, where ... is all the permissions required, is good enough. I don't really care which one is missing, I want to know what permissions I need so I can set those.

Reply:
As a user I would honestly want to know which ones are missing, not which are in general required. But that is more a question of taste rather than objective superiority I think. We can simply agree upon a single method and I will implement it that way.

Reply:
I changed it to @kodebach suggestion